src/gui/ide.ts

```diff
@@ -0,0 +1,2670 @@
+			bullet.style.cursor = "pointer";
+			bullet.addEventListener("click", function(event)
+					{
+						showdoc(href);
+						return false;
+					});
+		}
+		let td = tgui.createElement({"type": "td", "parent": tr, "classname": "ide"});
+		let lines = text.split('\n');
+		for (let i=0; i<lines.length; i++)
+		{
+			let s = lines[i];
+			let msg = tgui.createElement({"type": "div", "parent": td, "classname": "ide ide-message" + (type != "print" ? " ide-errormessage" : ""), "text": s});
+			if (typeof line !== 'undefined')
+			{
+				msg.ide_line = line;
+				msg.ide_ch = ch;
+				msg.addEventListener("click", function(event)
+						{
+							setCursorPosition(event.target.ide_line, event.target.ide_ch);
+							if (module.interpreter && (module.interpreter.status != "running" || !module.interpreter.background))
+							{
+								updateControls();
+							}
+							return false;
+						});
+			}
+		}
+		module.messagecontainer.scrollTop = module.messagecontainer.scrollHeight;
+		if (href) module.sourcecode.focus();
+	}
+	
+	// Stop the interpreter and clear all output,
+	// put the IDE into "not yet checked" mode.
+	function clear()
+	{
+		if (module.interpreter) module.interpreter.stopthread();
+		module.interpreter = null;
+	
+		tgui.clearElement(module.messages);
+		{
+			let ctx = module.turtle.getContext("2d");
+			ctx.setTransform(1, 0, 0, 1, 0, 0);
+			ctx.clearRect(0, 0, module.turtle.width, module.turtle.height);
+		}
+		if (module.interpreter) module.interpreter.service.turtle.reset.call(module.interpreter, 0, 0, 0, true);
+	
+		let ctx = module.canvas.getContext("2d");
+		ctx.setTransform(1, 0, 0, 1, 0, 0);
+		ctx.clearRect(0, 0, module.canvas.width, module.canvas.height);
+		ctx.lineWidth = 1;
+		ctx.fillStyle = "#000";
+		ctx.strokeStyle = "#000";
+		ctx.font = "16px Helvetica";
+		ctx.textAlign = "left";
+		ctx.textBaseline = "top";
+	}
+	
+	// Prepare everything for the program to start running,
+	// put the IDE into stepping mode at the start of the program.
+	module.prepare_run = function()
+	{
+		clear();
+	
+		// make sure that there is a trailing line for the "end" breakpoint
+		let source = module.sourcecode.getValue();
+		if (source.length != 0 && source[source.length - 1] != '\n')
+		{
+			source += '\n';
+			module.sourcecode.getDoc().replaceRange('\n', CodeMirror.Pos(module.sourcecode.lastLine()));
+		}
+	
+		let result = Parser.parse(source, options);
+		let program = result.program;
+		let html = "";
+		let errors = result.errors;
+		if (errors)
+		{
+			for (let i=0; i<errors.length; i++)
+			{
+				let err = errors[i];
+				module.addMessage(err.type, err.type + " in line " + err.line + ": " + err.message, err.line, err.ch, err.href);
+			}
+		}
+	
+		if (program)
+		{
+			module.interpreter = new Interpreter(program, defaultService);
+			module.interpreter.service.documentation_mode = false;
+			module.interpreter.localStorage = window.localStorage;
+			module.interpreter.service.audioContext = new AudioContext();
```

Comment:
```suggestion
```
This can be safely removed. `audioContext` is already set in `defaultService.ts` ([here](https://github.com/TheHllm/tscript/blob/7ce9751006b088607458d5ede32757d34d6aa7f8/src/interpreter/defaultService.ts#L266)). Also, accessing `AudioContext` breaks the IDE in Safari.